nixos/modules/config/system-path.nix

```diff
@@ -7,6 +7,118 @@ with lib;
 
 let
 
+  # This type can accept three different types of expressions:
+  #   * A list of packages like `[ pkgs.foo pkgs.bar ]`
+  #   * An attribute set of packages like `{ inherit (pkgs) foo bar; }`
+  #   * A more structured attribute set of packages like `{ "foo".package = pkgs.foo.override { hasBar = false; }; }`
+  #
+  # The first two cases will be coerced into the third one, which allows the
+  # package to be disabled and its package and list of outputs to be overridden.
+  packagesType =
+    let
+      inherit (lib) types;
+
+      outputsFor = drv:
+        if drv.outputSpecified or false
+        then [ drv.outputName ]
+        else ((drv.meta or {}).outputsToInstall or [ (drv.outputName or "out") ]);
+
+      mkOutputs =
+        let
+          mkEntry = output: lib.nameValuePair output true;
+        in
+        outputs: lib.listToAttrs (map mkEntry outputs);
+
+      drvToAttrs = drv:
+        let
+          defaultOutput = (drv.${(lib.head (drv.outputs or [ "out" ]))}
+            or (throw "something is up with the outputs of this derivation: ${drv}")) // {
+              # We want the generic derivation without outputs specified so that
+              # we can set `meta.outputsToInstall` later.
+              outputSpecified = false;
+            };
+        in
+        {
+          package = defaultOutput;
+          outputs = mkOutputs (outputsFor drv);
+        };
+
+      drvToAttrsFolder = acc: drv: lib.recursiveUpdate acc {
+        "${builtins.unsafeDiscardStringContext drv.drvPath}" = drvToAttrs drv;
+      };
+
+      pkgsToAttrs = lib.foldl' drvToAttrsFolder {};
```

Comment:
`recursiveUpdate` is dangerous because it silently overrides values (which indeed can happen here). It shouldn't be necessary if you know the structure. In this case something like this should work:

```nix
pkgsToAttrs = list:
  lib.mapAttrs (name: values:
    if length values > 1 then
      ?
    else
      head values
  ) (lib.groupBy (drv:
    builtins.unsafeDiscardStringContext drv.drvPath
  ) list
```

Reply:
Alternatively to @roberth's idea of using the `drvPath` as the attribute name, we could also generate attribute names from the list index, such that you can't even get a stable index to override. This would at least avoid the somewhat ugly `unsafeDiscardStringContext` and attribute names.